browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+  const args = {
+    label_1: getModel().foreground,
+    label_2: getModel().background
+  };
+  const action = new BackendAction('replace', args);
+  addFencedAction(action);
+};
+
+const watershed = (context) => {
+  const args = {
+    frame: getModel().frame,
+    label: context.storedLabel,
+    x1_location: context.storedX,
+    y1_location: context.storedY,
+    x2_location: getCanvas().imgX,
+    y2_location: getCanvas().imgY
+  };
+  const action = new BackendAction('watershed', args);
+  addFencedAction(action);
+};
+
+// tool states
+const paintState = {
+  initial: 'idle',
+  entry: assign({ trace: [] }),
+  states: {
+    idle: {
+      on: {
+        mousedown: {
+          // prevents select and paint at the same time
+          cond: (_, event) => !event.shiftKey,
+          target: 'dragging',
+          actions: 'addToTrace'
+        }
+      }
+    },
+    dragging: {
+      on: {
+        mousemove: { actions: 'addToTrace' },
+        mouseup: { actions: 'draw' },
+      }
+    }
+  }
+};
+
+const thresholdState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        mousedown: {
+          target: 'dragging',
+          actions: 'storeClick'
+        }
+      }
+    },
+    dragging: {
+      on: {
+        mouseup: [
+          {
+            cond: 'nonemptyBox',
+            actions: 'threshold'
+          },
+          { target: 'idle' },
+        ]
+      }
+    }
+  },
+  on: {
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+const floodState = {
+  on: {
+    click: { actions: 'flood' }
+  }
+};
+
+const trimState = {
+  on: {
+    click: { actions: 'trim' }
+  }
+};
+
+const erodeDilateState = {
+  on: {
+    mousedown: {
+      actions: choose([
+        {
+          cond: 'leftMouse',
+          actions: 'erode'
+        },
+        {
+          cond: 'rightMouse',
+          actions: 'dilate'
+        }
+      ])
+    }
+  }
+};
+
+const autofitState = {
+  on: {
+    click: { actions: 'autofit' },
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+const watershedState = {
+  entry: 'storeClick',
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        click: {
+          cond: 'validSeed',
+          target: 'clicked',
+          actions: 'storeClick'
+        }
+      }
+    },
+    clicked: {
+      on: {
+        click: {
+          cond: 'validSecondSeed',
+          actions: 'watershed'
+        }
+      }
+    }
+  },
+  on: {
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles which current tool for mouse behavior.
+ */
+const toolbarState = {
+  initial: 'paint',
+  states: {
+    flood: floodState,
+    trim: trimState,
+    erodeDilate: erodeDilateState,
+    autofit: autofitState,
+    paint: paintState,
+    threshold: thresholdState,
+    watershed: watershedState,
+    hist: { type: 'history' }
```

Comment:
What does this `type` mean?

Reply:
This creates a (history state)[https://xstate.js.org/docs/guides/history.html]. This allows us to return to using the same tool once the previous action resolves, so we can use the same tool repeatedly.